components/layout/traversal.rs

```diff
@@ -185,8 +185,8 @@ impl<'a> PreorderDomTraversal for RecalcStyleForNode<'a> {
 
                     if node.as_element().is_some() {
                         // Perform the CSS selector matching.
-                        let stylist = unsafe { &*self.layout_context.shared.stylist };
-                        node.match_node(stylist,
+                        let stylist = self.layout_context.shared.stylist.read().unwrap();
```

Comment:
I don't see how that's possible...

Actually, I'm pretty sure the current usage of Stylist on master is actually threadsafe without any additional locking.  It's just impossible to prove that to the compiler at the moment because we can't attach a lifetime to SharedLayoutContext with the current WorkQueue API.